### PR TITLE
Rust Cross Compilation (1.32.0)

### PR DESCRIPTION
**Pull request by illegalprime**
###### Motivation for this change

This is the same as #50866 but re-based and updated for Rust 1.32.0. It also does some other extra stuff like moving all output dirs to `target/{release, debug}` and being able to specify the build mode (`release` or `debug`).

the following derivations build:

- [x] `pkgs.pkgsCross.aarch64-multiplatform.rustc`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.cargo`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.fd`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.synapse-bt`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.webmetro`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.miniserve`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.ripgrep`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.ethabi`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.zola`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.todiff`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.gitAndTools.git-ignore`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.gitAndTools.git-codeowners`
- [x] `pkgs.pkgsCross.aarch64-multiplatform.gitAndTools.git-absorb`
- [x] `pkgs.rustc` 

I will attempt to updated this as soon as I figure out how to build more derivations. There are a lot more rust packages in nixpkgs, but these are likely a good starting point.

Any edits to or help with these derivations is appreciated.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by illegalprime** [comment]
not sure why, maybe @matthewbauer might know, but I get these errors on a lot of rust packages: `ripgrep`, `zola`, etc.
```
/nix/store/hxwlbcm8bbhdacayc90jljcd76w434r0-aarch64-unknown-linux-gnu-binutils-2.31.1/bin/aarch64-unknown-linux-gnu-ld: /nix/store/f6lrlfbg9ddvvw20j6av4ym7hxy6kxx1-rustc-1.32.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-e27907af0047d202.rlib(std-e27907af0047d202.std.3x3nkc5s-cgu.0.rcgu.o): Relocations in generic ELF (EM: 0)
/nix/store/hxwlbcm8bbhdacayc90jljcd76w434r0-aarch64-unknown-linux-gnu-binutils-2.31.1/bin/aarch64-unknown-linux-gnu-ld: /nix/store/f6lrlfbg9ddvvw20j6av4ym7hxy6kxx1-rustc-1.32.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-e27907af0047d202.rlib: error adding symbols: file in wrong format
          collect2: error: ld returned 1 exit status
```

it seems like `libstd` is in the wrong format, even though `rust` seems to think it's right, (because it placed it in the `aarch64-unknown-linux-gnu` folder). so maybe `rustc` is being build incorrectly.

but the `fd` package seems to build fine, it's the exception however. i believe it's not using the same `libstd` somehow, but i know it's using the same `rustc`

**Comment by Mic92** [comment]
I have not looked yet at the `rlib` format, but I read it is a elf file with some extras, so using `readelf` should provide you some information.

**Comment by illegalprime** [comment]
@Mic92 yes, (see list above) cross-compiled `cargo`, `rustc`, and other packages I've tested are working! The only thing that isn't "working" is compiling `rustc` to be a cross-compiler that runs on the build platform and has cross-compiled host libs. I get around it with the hack I mentioned earlier.

the right architecture libs seems to get compiled, but there's something wrong with them. :/

**Comment by Mic92** [comment]
Can you upload? /nix/store/qzpw6c9i8jk3lq3wbaspawc7mcwslryz-my-rustc/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-e27907af0047d202.rlib

File: pkgs/applications/networking/p2p/synapse-bt/default.nix
Comment (Mic92) [inline pkgs/applications/networking/p2p/synapse-bt/default.nix]: Is this something we should set by default in `buildRustPackage`?

Comment (illegalprime) [reply pkgs/applications/networking/p2p/synapse-bt/default.nix]: possibly, it's for `pkg-config` so ostensibly it would need to be set anywhere `pkg-config` is used during cross-compilation. Don't know how to handle all those cases though.

**Comment by illegalprime** [comment]
@Mic92 so this happens when you build `rustc` with `buildPlatform == hostPlatform != targetPlatform`, here you go:
[libstd-e27907af0047d202.zip](https://github.com/NixOS/nixpkgs/files/2957999/libstd-e27907af0047d202.zip)

**Comment by Mic92** [comment]
Looks actually sane on the first glance:

```console
$ 7z x libstd-e27907af0047d202.zip
$ ar x libstd-e27907af0047d202.zip
$ file *.o
alloc.o:                                        ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
backtrace.o:                                    ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
dwarf.o:                                        ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
elf.o:                                          ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
fileline.o:                                     ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
posix.o:                                        ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
read.o:                                         ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
sort.o:                                         ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
state.o:                                        ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.0.rcgu.o: ELF 64-bit LSB relocatable, no machine, version 1 (SYSV), not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.1.rcgu.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.2.rcgu.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.3.rcgu.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.4.rcgu.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.5.rcgu.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.6.rcgu.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.7.rcgu.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.8.rcgu.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
std-e27907af0047d202.std.3x3nkc5s-cgu.9.rcgu.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), with debug_info, not stripped
```

**Comment by illegalprime** [comment]
@Mic92 yea, could setting the Machine correctly really be all that's needed?

**Comment by Mic92** [comment]
@illegalprime could be. I have not figured out yet, how I can run gdb on the build compiler otherwise I would set a breakpoint in the right location. The hack is ok right now, also annoying because it forces cross-compilers to build a rustc they probably never need. Maybe this can be reported upstream though. I wonder how they build those stdlibs.

File: pkgs/applications/networking/p2p/synapse-bt/default.nix
Comment (matthewbauer) [reply pkgs/applications/networking/p2p/synapse-bt/default.nix]: Actually it is a pkgconfig-rs thing i think

File: pkgs/development/compilers/rust/rustc.nix
Comment (matthewbauer) [inline pkgs/development/compilers/rust/rustc.nix]: This looks wrong. /cc @ericson2314

Comment (matthewbauer) [reply pkgs/development/compilers/rust/rustc.nix]: Actually it might be correct just a weird thing to set. You can’t run the target compiler on your building machine

File: pkgs/applications/networking/p2p/synapse-bt/default.nix
Comment (Mic92) [reply pkgs/applications/networking/p2p/synapse-bt/default.nix]: Maybe some hook in buildRustPackage that checks if pkg-config is available and sets `PKG_CONFIG_ALLOW_CROSS` in that case?

File: pkgs/build-support/rust/default.nix
Comment (illegalprime) [reply pkgs/build-support/rust/default.nix]: if i try to use `targetPackages.rustc` I get:
`error: attribute 'rustc' missing,`

Comment (Ericson2314) [reply pkgs/build-support/rust/default.nix]: Shouldn't it be `buildPackages`?

Comment (illegalprime) [reply pkgs/build-support/rust/default.nix]: this `rustc` is supposed to be the `rustc` whose host & target is aarch64

Comment (illegalprime) [reply pkgs/build-support/rust/default.nix]: I think `pkgs.rustc` is where just the target is aarch64 and build & host are x86

**Comment by illegalprime** [comment]
turns out it was the stripping in the fixupPhase that was creating the bad rlibs, not rust's bootstrapping process, :(

**Comment by Mic92** [comment]
Good news. Anything left to do then?

**Comment by illegalprime** [comment]
@Mic92 i have an interest in cross compiling for 32-bit arm (armv7), doing:
```
nix build -f . pkgsCross.armv7l-hf-multiplatform.ripgrep
```
breaks because Rust doesn't recognize the triple `armv7l-unknown-linux-gnueabihf`, but is _does_ recognize `armv7-unknown-linux-gnueabihf` (no l). my whole wanting rust-cross-compilation is for this target so I'd like to be able to get that working before this merges, should I:

1. somewhere within buildRustPackage take that triple and remove the `l` when giving it to rust
2. add a new entry in `pkgsCross`
3. ???

if you think no. 2 is preferable, how do I do that? the entire layout of `pkgsCross` is a mystery to me

**Comment by illegalprime** [comment]
another thing since we're here, how do people feel about https://www.hadean.com/blog/managing-rust-dependencies-with-nix-part-ii (i'm not the author, or know the author)? It seems more nix-ish than the current setup, but it didn't seem very popular in #my other main goal is to be able to support rust builds where some crates are in directories

**Comment by Ericson2314** [comment]
I rebased @Mic92's old stuff and changed some things. I'll work to get that rebased so I can combine with this.

**Comment by illegalprime** [comment]
need 2 do another thing: need to add the `.cargo/config` in the sourceRoot instead of in pwd during configure.

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nixpkgs/pull/57611/files is something that would make this easier.

**Comment by Ericson2314** [comment]
https://github.com/Ericson2314/nixpkgs/branche/rust-cross here my old modification, rebased to have the same branch as this, but not yet tested). However, if #57611 is merged, we should rebase both onto that before combining.

**Comment by illegalprime** [comment]
After some hacking, ignore my previous 2 comments, I'm all good on this PR I think it's ready to merge! :tada:

File: pkgs/development/compilers/rust/rustc.nix
Comment (illegalprime) [inline pkgs/development/compilers/rust/rustc.nix]: this might be wrong, need to test

Comment (illegalprime) [reply pkgs/development/compilers/rust/rustc.nix]: so when building under ARMv7 rust tries to call `llvm-config` to get the libs to link against. `buildLlvmShared` has an `llvm-config` that can run on the build system but the wrong shared libraries, while `llvm-config` is the opposite (wrong binary, right libs). Not sure why this problem doesn't show up on AArch64.

Comment (illegalprime) [reply pkgs/development/compilers/rust/rustc.nix]: Maybe a shim `llvm-config` script should be introduced, this is similar to meta-amd's: https://git.yoctoproject.org/cgit/cgit.cgi/meta-amd/tree/common/recipes-core/llvm/files/0001-llvm-config-allow-overriding-libdir-through-cmdline.patch?h=sumo

Comment (Mic92) [reply pkgs/development/compilers/rust/rustc.nix]: Since there should be no arch-specific one could wrap around the llvm-config
and then just do `sed -e "s!${buildPackages.llvm}!${llvm}!" on the output?

File: pkgs/development/compilers/rust/rustc.nix
Comment (Mic92) [inline pkgs/development/compilers/rust/rustc.nix]: Ah. Exactly what I proposed! I think we should make this part of llvm itself.
As it can be useful in other places.

Comment (illegalprime) [reply pkgs/development/compilers/rust/rustc.nix]: would I add something like `wrapLlvmConfig` to the `llvm` directory in `pkgs/development/compilers/llvm` and then add it to `pkgs/top-level/all-packages.nix`? or is there a more conventional way to add a 'global function' rather than a package?

Comment (Mic92) [reply pkgs/development/compilers/rust/rustc.nix]: I am not sure about the best way to deal with this. Ideally, users who are not aware of cross-compiling should not have to take care of this when adding llvm to a project. @Ericson2314 how do you dealt with this in the past?

**Comment by Laaas** [comment]
Will this allow cross-compiling to Windows targets?

**Comment by Mic92** [comment]
@Laaas 	I have not tested that, but I think it could already work or would just need some other smaller modifications. It was not too much effort to get go applications cross-compiled for the windows back then: https://github.com/NixOS/nixpkgs/pull/50928

**Comment by Mic92** [comment]
@illegalprime there is nothing left on your TODO list, right?

**Comment by illegalprime** [comment]
@Mic92 this could probably go in, the only things left would be:
1. maybe put the `llvm-config` shim somewhere else so other packages might use it (https://github.com/NixOS/nixpkgs/pull/56540#discussion_r267405691), don't know what to do here
2. map `armv7l -> armv7` in the rust ecosystem to be able to build `pkgsCross.armv7l-hf-multiplatform` but maybe that should go in another PR? I'm not super sure how i'd do the mapping.

**Comment by Ericson2314** [comment]
I'm very close on https://github.com/NixOS/nixpkgs/pull/57611/files (just need some final review of manual changes). Would you mind waiting a tiny bit longer for that to be done, and then I'll rebase this?

**Comment by illegalprime** [comment]
@Mic92 NP, do you have documented somewhere the philosophy and plan for nixpkgs cross-comp? things like `depsTargetTargetPropagated` and that PR make me feel like there's a larger architecture goal at play that I don't quite understand.

**Comment by Ericson2314** [comment]
I assume you meant me? :). So https://nixos.org/nixpkgs/manual/#ssec-stdenv-dependencies is the main section describing the dependencies and how they fit together, but in that PR I extended the corresponding section the cross compilation section. Hopefully https://hydra.nixos.org/build/91228503# has it in a moment.

**Comment by Ericson2314** [comment]
OK now that that #57611 is merged:
- https://github.com/Ericson2314/nixpkgs/commits/rust-cross is the rebased version of mine
- https://github.com/Ericson2314/nixpkgs/commits/rust-cross-2 is the rebased version of this, with one squash.

The first thing to note is @Mic92 did take my changes (all-packages.nix changes were moved to an I guess orphaned `default.nix`) I missed that before, sorry. This means there will be little-to-nothing to double check against my branch :).

The second things is @illegalprime your PR is harder to review because some commits "undo" other commits (revert or moral revert). Could you `git rebase -i` so the commits tell a nicer, if artificial, "story"? (Feel free to start with the rebase I did, which squashed the "remove jemalloc" change since the original commit added it back only after an prior rebase's conflict resolution.)

After that's done I'll be able to better wrap my head around everything, and also set about incorporating the new `pkgsFooBar` ewnly-added in #57611, which will help with cross compiling rustc (build != host == target).

**Comment by illegalprime** [comment]
thanks for the links to the docs @Mic92 , i rebased it, we'll see how the tests go

**Review by Ericson2314 — APPROVED** [review]
Thanks again @illegalprime and @Mic92! I pruned the remaining silly states from the history, and incorporated the new `pkgsBuild*`, so this is ready to go!

**Comment by Ericson2314** [comment]
https://github.com/Ericson2314/nixpkgs/commits/rust-cross-2 has the beginnings of a `buildRustCrate`, but that can be merged later.

**Comment by Ericson2314** [comment]
=NixOS/ofborg looked good enough to me, haha.

**Pull request merged**
